File: 116*/Solution.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if root is None or root.left is None:
            return
        
        start = root;
        root.left.next = root.right
        while root.next is not None:
            root.right.next = root.next.left
            root = root.next
            if root.left is not None:
                root.left.next = root.right
        
        self.connect(start.left)

File: 116*/Solution.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        if (root == null || root.left == null)
            return;
        
        TreeLinkNode start = root;
        root.left.next = root.right;
        while (root.next != null) {
            root.right.next = root.next.left;
            root = root.next;
            if (root.left != null) 
                root.left.next = root.right;
        }
        connect(start.left);
    }
}